### PR TITLE
Molecule: add optional Hill order

### LINKED ISSUE #206: Should molecular formulas be Hill-ordered?
**Describe the bug**
Molecular formulas (`Molecule.get_molecular_formula`) are currently alphabetically ordered. Should they be Hill ordered?

**Additional context**
This would require a data migration in Fractal. 

Hill order: the number of carbon atoms in a molecule is indicated first, the number of hydrogen atoms next, and then the number of all other chemical elements subsequently, in alphabetical order of the chemical symbols. When the formula contains no carbon, all the elements, including hydrogen, are listed alphabetically.

### PR DESCRIPTION
**Pull request by mattwelborn**
## Description
This PR adds the option to Hill-order molecular formulas. It also adds `Molecule.order_molecular_formula` to re-order existing formulas (needed by Fractal). 

Closes #206.

## Changelog description
Added the option to Hill-order molecular formulas. 

## Status

- [x] Code base linted
- [x] Ready to go

### CODE REVIEW

File: qcelemental/models/molecule.py [hunk 1]
```diff
@@ -753,17 +758,77 @@ def get_molecular_formula(self):
...
+            List of chemical symbols
+        order: str, optional
+            sorting order of the formula. Valid choices are "alphabetical" and "hill".
+
+        Returns
+        -------
+        str
+            The molecular formula
+        """
+
```
Comment (dgasmith) [inline qcelemental/models/molecule.py]: `order = order.lower()`

File: qcelemental/models/molecule.py [hunk 2]
```diff
@@ -753,17 +758,77 @@ def get_molecular_formula(self):
...
+        for k in element_order:
             c = count[k]
             ret.append(k)
             if c > 1:
                 ret.append(str(c))
 
         return "".join(ret)
 
+    @staticmethod
+    def order_molecular_formula(formula: str, order: str = "alphabetical") -> str:
```
Comment (dgasmith) [inline qcelemental/models/molecule.py]: I worry a bit about function explosion with items like this on the Molecule class. Can you think of another place to put it?

File: qcelemental/tests/test_molecule.py [hunk 3]
```diff
@@ -47,6 +47,34 @@ def test_molecule_data_constructor_numpy():
...
+        ("CCl4", "alphabetical", "CCl4"),
+        ("CCl4", "hill", "CCl4"),
+        ("CBr4", "alphabetical", "Br4C"),
+        ("CBr4", "hill", "CBr4"),
+        ("CBrH3", "alphabetical", "BrCH3"),
+        ("CBrH3", "hill", "CH3Br"),
+    ],
+)
+def test_order_molecular_formula(input, order, expected):
+    assert Molecule.order_molecular_formula(input, order) == expected
```
Comment (dgasmith) [inline qcelemental/tests/test_molecule.py]: `order=order`

File: qcelemental/models/molecule.py [hunk 4]
```diff
@@ -727,10 +728,14 @@ def get_hash(self):
...
+    def get_molecular_formula(self, order: str = "alphabetical") -> str:
         """
-        Returns the molecular formula for a molecule. Atom symbols are sorted from
-        A-Z.
+        Returns the molecular formula for a molecule.
+
+        Pararmeters
+        -----------
+        order: str, optional
+            sorting order of the formula. Valid choices are "alphabetical" and "hill".
```
Comment (dgasmith) [inline qcelemental/models/molecule.py]: "Sorting order", ditto below.

File: qcelemental/models/molecule.py [hunk 5]
```diff
@@ -753,17 +758,77 @@ def get_molecular_formula(self):
...
+        -----------
+        symbols: List[str]
+            List of chemical symbols
+        order: str, optional
+            sorting order of the formula. Valid choices are "alphabetical" and "hill".
+
+        Returns
+        -------
+        str
+            The molecular formula
```
Comment (dgasmith) [inline qcelemental/models/molecule.py]: "The molecular formula."

File: qcelemental/models/molecule.py [hunk 6]
```diff
@@ -753,17 +758,77 @@ def get_molecular_formula(self):
...
+        count = collections.defaultdict(int)
+        for match in matches:
+            match_n = re.match("(\D+)(\d*)", match)
+            assert match_n
+            if match_n.group(2) == "":
+                n = 1
+            else:
+                n = int(match_n.group(2))
+            count[match_n.group(1)] += n
+        symbols = [k for k, v in count.items() for i in range(v)]
```
Comment (dgasmith) [inline qcelemental/models/molecule.py]: We should probably `symbol.title()` to make sure things don't slip through like `ch4`.

Comment (mattwelborn) [reply qcelemental/models/molecule.py]: The regex won't match symbols that aren't title case. We don't know if "ch4" should be "CH4" or "Ch4". I've added a check for malformed molecular formulae.

File: qcelemental/models/molecule.py (hunk 2 above)
Comment (mattwelborn) [reply qcelemental/models/molecule.py]: molutil???

Comment (dgasmith) [reply qcelemental/models/molecule.py]: Yea, would that be ok?

Comment (mattwelborn) [reply qcelemental/models/molecule.py]: Yep!

**Pull request merged**